cmd/kubeadm/app/cmd/phases/certs.go

```diff
@@ -188,51 +186,52 @@ func getSANDescription(certSpec *certsphase.KubeadmCert) string {
 	return fmt.Sprintf("\n\nDefault SANs are %s", strings.Join(sans, ", "))
 }
 
-func addFlags(cmd *cobra.Command, cfgPath *string, cfg *kubeadmapiv1beta1.InitConfiguration, addAPIFlags bool) {
-	options.AddCertificateDirFlag(cmd.Flags(), &cfg.CertificatesDir)
-	options.AddConfigFlag(cmd.Flags(), cfgPath)
-	if addAPIFlags {
-		cmd.Flags().StringVar(&cfg.Networking.DNSDomain, "service-dns-domain", cfg.Networking.DNSDomain, "Alternative domain for services, to use for the API server serving cert")
-		cmd.Flags().StringVar(&cfg.Networking.ServiceSubnet, "service-cidr", cfg.Networking.ServiceSubnet, "Alternative range of IP address for service VIPs, from which derives the internal API server VIP that will be added to the API Server serving cert")
-		cmd.Flags().StringSliceVar(&cfg.APIServerCertSANs, "apiserver-cert-extra-sans", []string{}, "Optional extra altnames to use for the API server serving cert. Can be both IP addresses and DNS names")
-		cmd.Flags().StringVar(&cfg.APIEndpoint.AdvertiseAddress, "apiserver-advertise-address", cfg.APIEndpoint.AdvertiseAddress, "The IP address the API server is accessible on, to use for the API server serving cert")
-	}
-}
+func makePhasesForCA(ca *certsphase.KubeadmCert, certList []*certsphase.KubeadmCert) []workflow.Phase {
+	subPhases := []workflow.Phase{}
 
-func makeCommandsForCA(ca *certsphase.KubeadmCert, certList []*certsphase.KubeadmCert, cfgPath *string, cfg *kubeadmapiv1beta1.InitConfiguration) []*cobra.Command {
-	subCmds := []*cobra.Command{}
+	caPhase := makePhase(ca)
+	caPhase.Run = func(c workflow.RunData) error {
+		data, ok := c.(certsData)
+		if !ok {
+			return errors.New("certs phase invoked with an invalid data struct")
+		}
 
-	caCmd := makeCmd(ca, cfgPath, cfg)
-	caCmd.Run = func(cmd *cobra.Command, args []string) {
-		internalcfg, err := configutil.ConfigFileAndDefaultsToInternalConfig(*cfgPath, cfg)
-		kubeadmutil.CheckErr(err)
+		if err := certsphase.CreateCACertAndKeyFiles(ca, data.Cfg()); err != nil {
+			return err
+		}
 
-		err = certsphase.CreateCACertAndKeyFiles(ca, internalcfg)
-		kubeadmutil.CheckErr(err)
+		return nil
 	}
 
-	subCmds = append(subCmds, caCmd)
+	subPhases = append(subPhases, caPhase)
 
 	for _, cert := range certList {
-		certCmd := makeCommandForCert(cert, ca, cfgPath, cfg)
-		subCmds = append(subCmds, certCmd)
+		certPhase := makePhaseForCert(cert, ca)
+		subPhases = append(subPhases, certPhase)
 	}
 
-	return subCmds
+	return subPhases
 }
 
-func makeCommandForCert(cert *certsphase.KubeadmCert, caCert *certsphase.KubeadmCert, cfgPath *string, cfg *kubeadmapiv1beta1.InitConfiguration) *cobra.Command {
-	certCmd := makeCmd(cert, cfgPath, cfg)
+func makePhaseForCert(cert *certsphase.KubeadmCert, caCert *certsphase.KubeadmCert) workflow.Phase {
+	certPhase := makePhase(cert)
+
+	certPhase.Run = func(c workflow.RunData) error {
```

Comment:
Same as above: move into a separated func, handle external ca and certificateDir for dryrun

Reply:
Updated separate func, have to handle the external ca and certificateDir yet